# async, await

- async

  - 프로미스 체이닝보다 가독성이 좋아짐

  ```
  //async 이용
  async function getName(){
      return "Mike";
      //return Promise.resolve("Mike");

      //에러 처리 반환일 때
      //throw new Error("err..");
  }

  //getName() 함수는 Promise 반환
  get.Name().then(name => { //catch로 받으면 에러 반환 가능
      console.log(name); // Mike 출력
  })
  ```

- await

  - await 키워드 옆에는 프로미스가 오고 그 프로미스가 처리될 때까지 기다림
  - await은 async 함수 내부에서만 사용

  ```
  //async, await
  function getName(name){
      return new Promise((resolve,reject) => {
          setTimeout(() => {
              resolve(name);
          }, 1000);
      });
  }

  //await으로 getName() resolve 기다림.
  async function showName(){
      const result = await getName("Mike");
      console.log(result); // 1초 후에 Mike 찍힘
  }
  ```

  ```
  const f1 = () => {
      return new Promise((res,rej) => {
          setTimeout(() =>{
              res("1번 주문 완료");
          },1000);
      });
  };

  const f2 = (message) => {
      console.log(message);
      return new Promise((res,rej) => {
          setTimeout(() =>{
              res("2번 주문 완료");
          },3000);
      });
  };

  const f3 = (message) => {
      console.log(message);
      return new Promise((res,rej) => {
          setTimeout(() =>{
              res("3번 주문 완료");
          },2000);
      });
  };

  console.log("시작");
  async function order(){
      try{
          const result1 = await f1();
          const result2 = await f2(result1);
          const result3 = await f3(result2);
          console.log(result3);
          //Promise.all() 사용하면
          //const result = await Promise.all([f1(),f2(),f3()]);
          //console.log(result);
      } catch(e){
          console.log(e);
      }
      console.log("끝");
  }

  order();
  ```

  - try - catch 문으로 try 시도하고 에러 발생하면 catch 부분으로 넘어간 후 마지막 console.log("끝") 실행
  - 예를 들어 중간에 f2() 함수가 rej로 error를 반환한다면 시작 -> 1번 주문 완료 -> Error -> 끝

# Generator

- 함수의 실행을 멈췄다가 재개할 수 있는 기능
- function\*, yield 키워드 사용
- next(), return(), throw()
- generator 객체가 반환되고 변수에 함수를 할당했을 때 함수명.next(); 하면 가장 가까운 yield까지 실행되고 데이터 객체를 반환
  - 여기서 데이터 객체는
    - value: yield 오른쪽 값, 마지막 return(없으면 undefined)
    - done: 함수가 끝났으면 treu, 아니면 false
- iterable, iterator
  - iterable : 반복 가능한 객체, Symbol.iterator 메서드가 있고 이는 iterator를 반환해야 한다.
  - iterator : 시퀀스를 정의하고 종료시의 반환값을 잠재적으로 정의하는 객체, value와 done이란 프로퍼티를 갖는 객체를 반환하는 객체 내 메서드로 next가 있어야함.
- 값을 미리 만들어두지 않아 메모리 관리 측면에서 효율적(break 없는 무한루프 while문도 가능)
- 이용 예시 : Redux Saga

```
function* fn(){
    try{
        console.log(1);
        yield 1;
        console.log(2);
        yield 2;
        console.log(3);
        console.log(4);
        yield 3;
        return "finish";
    } catch(e){
        console.log(e);
    }
}

const a = fn();

a[Symbol.iterator]() === a; // true
```

- next()로 진행하다가 return('end'); 하게 되면 {value: 'end', done: true}
- throw(new Error("err")); 하게 되면 {value : undefined, done: true}와 Error: err라는 에러 로그 찍힘

```
function* fn(){
    const num1 = yield "첫 번째 숫자 입력";
    console.log(num1);

    const num2 = yield "두 번째 숫자 입력";
    console.log(num2);

    return num1 + num2;
}

const a = fn();

//console 창에서
a.next(); // {value:첫번째 숫자입력, done:false}
a.next(2); // num1에 2 들어가면서 두번째 yield 실행
a.next(4); // num2에 4 들어가면서 2+4 = 6 반환
```

```
//generator 중첩, yield*
function* gen1(){
    yield "2";
    yield "3";
    yield "4";
    yield "5";
}

function* gen2(){
    yield "1";
    yield* gen1();
    yield "6";
}
```

//console 창에서
console.log(...gen2()); //done이 true가 될 때까지 펼쳐줌

### iterable, iterator

- 배열, 문자열도 iterable 하다

```
const arr = [1,2,3,4,5];
const it = arr[Symbol.iterator]();
//typeof 변수명[Symbol.iterator]; -> "function"

it.next(); // {value: 1, done: false};
it.next(); // {value: 2, done: false};

//for of 로 순회 가능
for(let num of arr){
    console.log(num);
}

function* fn() {
    yield 4;
    yield 5;
    yield 6;
}

const a = fn();

for(let num of a){
    console.log(num); //4 5 6 출력
}
```
